helm: default to cronJob as tls.auto.method

kaworu opened the pull request:
For both Hubble and Clustermesh. Deprecate the previous default `helm` method on the way.

See first task of https://github.com/cilium/cilium/issues/25345.

kaworu committed:
helm: deprecate the helm auto tls method

This method was introduced as a quick way to get TLS setup for Hubble
and clustermesh before other method were implemented. Because of its
inherent limitations (not gitops friendly, forcing certificate
re-generation on Helm ugprades) the default is going to be changed to
the cronJob method, and thus the helm method is now deprecated and
planned for removal in v1.16.

Signed-off-by: Alexandre Perrin <alex@isovalent.com>

kaworu committed:
helm: use cronJob as the default auto tls method

The previous commit deprecated the default helm method, and this commit
promote the cronJob method as the new default.

Signed-off-by: Alexandre Perrin <alex@isovalent.com>

rolinh reviewed (approved):
LGTM, thanks for tackling this!

gandro reviewed (approved):
I'm fine with the change as is, but if we make `cronJob` the default, we likely also need to figure out how `helm uninstall` is supposed to work in `cronJob` mode _before_ the feature freeze in a couple of weeks: https://github.com/cilium/cilium-cli/issues/1648

giorio94 commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -994,13 +994,15 @@ hubble:
+      # - cronJob:      This method uses a Kubernetes CronJob the generate
+      #                 any certificates not provided by the user at
+      #                 installation time.
+      # - certmanager:  This method use cert-manager to generate & rotate
```
```suggestion
      # - certmanager:  This method uses cert-manager to generate & rotate
```

Nit

giorio94 commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -2746,13 +2748,15 @@ clustermesh:
+        # - cronJob:      This method uses a Kubernetes CronJob the generate
+        #                 any certificates not provided by the user at
+        #                 installation time.
+        # - certmanager:  This method use cert-manager to generate & rotate
```
```suggestion
        # - certmanager:  This method uses cert-manager to generate & rotate
```

Nit

giorio94 commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -2771,7 +2771,7 @@ clustermesh:
         # fourth month. For syntax, see
         # https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/#schedule-syntax
-        # schedule: "0 0 1 */4 *"
+        schedule: "0 0 1 */4 *"
```
Nit. The comment above also need to be updated to reflect this change.

giorio94 reviewed (approved):
Thanks! I just left a couple of nits inline. I guess that you should be able to drop the first commit once https://github.com/cilium/cilium/pull/25261 gets merged.

zacharysarah commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -994,13 +994,15 @@ hubble:
       enabled: true
-      # -- Set the method to auto-generate certificates. Supported values:
+      # -- Sets the method to auto-generate certificates. Supported values:
+      # - cronJob:      This method uses a Kubernetes CronJob the generate
```
spelling nit
```suggestion
      # - cronJob:      This method uses a Kubernetes CronJob to generate
```

zacharysarah commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -2746,13 +2748,15 @@ clustermesh:
         enabled: true
-        # Sets the method to auto-generate certificates. Supported values:
+        # -- Sets the method to auto-generate certificates. Supported values:
+        # - cronJob:      This method uses a Kubernetes CronJob the generate
```
Spelling nit
```suggestion
        # - cronJob:      This method uses a Kubernetes CronJob to generate
```

zacharysarah reviewed (approved):
A couple of nits, otherwise LGTM. Approving with the understanding that changes must be made prior to merge.

michi-covalent reviewed (approved):
let's harmonize 🧘

kaworu committed:
doc: remove stale reference to hubble-ca-cert

The hubble-ca-cert ConfigMap was removed by 4635ffaced50578e7a443bd4efa7cee42ae52cb7.

Signed-off-by: Alexandre Perrin <alex@isovalent.com>

kaworu committed:
test: remove hubble-ca-secret from the cleanup list

All other references to this secret were removed in
fcc927f80756728dd0915d0091694433471c4fa2.

Signed-off-by: Alexandre Perrin <alex@isovalent.com>

joamaki commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -994,12 +994,14 @@ hubble:
       enabled: true
-      # -- Set the method to auto-generate certificates. Supported values:
+      # -- Sets the method to auto-generate certificates. Supported values:
+      # - cronJob:      This method uses a Kubernetes CronJob to generate any
```
nit: "cronjob" would match capitalization of "certmanager" and reads better

kaworu replied:
make sense, we could update the value here and do a case-insensitive check in the template for backward compatibility.

kaworu replied:
Thinking back about this we have `.Values.certgen.annotations.cronJob`, so we'll keep cronJob for now for consistency with this annotation Helm value.

viktor-kurchenko reviewed (approved):
LGTM.

giorio94 commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -1082,12 +1082,14 @@ hubble:
-      #                 certificates not provided by the user at installation
-      #                 time.
-      # - certmanager:  This method use cert-manager to generate & rotate certificates.
+      #                 Deprecated and will be removed in Cilium v1.15.
```
I guess this should be 1.16 now. The same below, and in the commit message.

kaworu replied:
Good catch thank you!

giorio94 reviewed (changes requested):
Overall /lgtm. Just one comment about the version in which the deprecated method will be removed.
